rclcpp/src/rclcpp/clock.cpp

```diff
@@ -182,6 +182,62 @@ Clock::sleep_for(Duration rel_time, Context::SharedPtr context)
   return sleep_until(now() + rel_time, context);
 }
 
+bool
+Clock::is_valid()
+{
+  switch (get_clock_type()) {
+    case RCL_ROS_TIME:
+    case RCL_STEADY_TIME:
+    case RCL_SYSTEM_TIME:
+      return now().nanoseconds() > 0;
+
+    case RCL_CLOCK_UNINITIALIZED:
+    default:
+      return false;
+  }
+}
+
+bool
+Clock::wait_for_valid(Context::SharedPtr context)
+{
+  if (get_clock_type() == RCL_CLOCK_UNINITIALIZED) {
+    throw std::runtime_error(
+            "clock cannot be waited on as its clock_type is RCL_CLOCK_UNINITIALIZED");
+  }
+
+  // Wait until the first valid time
+  return sleep_until(rclcpp::Time(0, 1, get_clock_type()), context);
+}
+
+bool
+Clock::wait_for_valid(
+  const Duration & timeout,
+  Context::SharedPtr context,
+  Duration wait_tick_ns)
+{
+  if (!context || !context->is_valid()) {
+    throw std::runtime_error("context cannot be slept with because it's invalid");
+  }
+
+  if (get_clock_type() == RCL_CLOCK_UNINITIALIZED) {
+    throw std::runtime_error(
+            "clock cannot be waited on as its clock_type is RCL_CLOCK_UNINITIALIZED");
+  }
+
+  Clock timeout_clock = Clock(RCL_SYSTEM_TIME);
+  Time start = timeout_clock.now();
+
+  while (!is_valid() && context->is_valid()) {
+    timeout_clock.sleep_for(Duration(wait_tick_ns));
+    if (timeout_clock.now() - start > timeout) {
+      return false;
+    }
+  }
+
+  return context->is_valid();
```

Comment:
should we call `is_valid` instead? i think we are not checking if the context is valid or invalid?

Reply:
Fair point, initially it was meant to check if we escaped the loop if the context became true or false, but it makes sense to just report the validity.

Likewise, I'll edit the `false` above to similarly report the validity.

Reply:
https://github.com/ros2/rclcpp/pull/2040/commits/a0091d0e5527fa45a09a77c70c98b8be82fe5b98